File: modules/tools/transcriber-service/internal/store/store.go
// Package store wraps the SQLite-backed job persistence.
package store

import (
	"database/sql"
	"fmt"
	"time"

	_ "modernc.org/sqlite"

	"github.com/akdasa-studios/lectorium/modules/tools/transcriber-service/internal/job"
)

const schema = `
CREATE TABLE IF NOT EXISTS jobs (
  job_id TEXT PRIMARY KEY,
  filename TEXT NOT NULL,
  language TEXT,
  status TEXT NOT NULL CHECK(status IN ('queued','running','done','failed')),
  uploaded_at INTEGER NOT NULL,
  started_at INTEGER,
  completed_at INTEGER,
  duration_seconds REAL,
  processing_time_seconds REAL,
  rtfx REAL,
  confidence REAL,
  error TEXT
);
CREATE INDEX IF NOT EXISTS idx_jobs_status ON jobs(status, uploaded_at);
`

// Store is a thin SQLite wrapper. Safe for concurrent use (SQLite serializes writes).
type Store struct {
	db *sql.DB
}

// Open opens (and migrates) the SQLite database at path.
func Open(path string) (*Store, error) {
	db, err := sql.Open("sqlite", path+"?_pragma=journal_mode(WAL)&_pragma=busy_timeout(5000)")
	if err != nil {
		return nil, fmt.Errorf("open sqlite: %w", err)
	}
	if _, err := db.Exec(schema); err != nil {
		db.Close()
		return nil, fmt.Errorf("migrate: %w", err)
	}
	return &Store{db: db}, nil
}

// Close releases the database handle.
func (s *Store) Close() error { return s.db.Close() }

// Insert creates a new queued job.
func (s *Store) Insert(j *job.Job) error {
	_, err := s.db.Exec(
		`INSERT INTO jobs (job_id, filename, language, status, uploaded_at)
		 VALUES (?, ?, ?, ?, ?)`,
		j.JobID, j.Filename, nullStr(j.Language), j.Status, j.UploadedAt,
	)
	return err
}

// MarkRunning transitions a job to running.
func (s *Store) MarkRunning(jobID string) error {
	_, err := s.db.Exec(
		`UPDATE jobs SET status='running', started_at=? WHERE job_id=? AND status IN ('queued','running')`,
		time.Now().UnixMilli(), jobID,
	)
	return err
}

// MarkDone transitions a job to done with the final metrics.
func (s *Store) MarkDone(jobID string, m job.Metrics) error {
	_, err := s.db.Exec(
		`UPDATE jobs
		 SET status='done', completed_at=?,
		     processing_time_seconds=?, duration_seconds=?, confidence=?, rtfx=?
		 WHERE job_id=?`,
		time.Now().UnixMilli(),
		m.ProcessingTimeSeconds, m.DurationSeconds, m.Confidence, m.RTFx,
		jobID,
	)
	return err
}

// MarkFailed transitions a job to failed.
func (s *Store) MarkFailed(jobID, errMsg string) error {
	_, err := s.db.Exec(
		`UPDATE jobs SET status='failed', completed_at=?, error=? WHERE job_id=?`,
		time.Now().UnixMilli(), errMsg, jobID,
	)
	return err
}

// ResetRunningToQueued is called at startup: any job in 'running' got there from
// a previous process that crashed mid-inference, so put it back in the queue.
func (s *Store) ResetRunningToQueued() (int, error) {
	res, err := s.db.Exec(
		`UPDATE jobs SET status='queued', started_at=NULL WHERE status='running'`,
	)
	if err != nil {
		return 0, err
	}
	n, _ := res.RowsAffected()
	return int(n), nil
}

// QueuedJobIDs returns the IDs of jobs currently in 'queued' state, oldest first.
// Used at startup to refeed pending work into fluidbatchd.
func (s *Store) QueuedJobIDs() ([]string, error) {
	rows, err := s.db.Query(`SELECT job_id FROM jobs WHERE status='queued' ORDER BY uploaded_at ASC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var ids []string
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return nil, err
		}
		ids = append(ids, id)
	}
	return ids, rows.Err()
}

// Get fetches one job by id. Returns nil, nil if not found.
func (s *Store) Get(jobID string) (*job.Job, error) {
	row := s.db.QueryRow(`
		SELECT job_id, filename, language, status, uploaded_at,
		       started_at, completed_at,
		       duration_seconds, processing_time_seconds, rtfx, confidence, error
		FROM jobs WHERE job_id=?`, jobID)
	return scanJob(row)
}

// List returns jobs filtered by status (empty = all), most recent first.
func (s *Store) List(status string, limit int) ([]*job.Job, error) {
	if limit <= 0 || limit > 1000 {
		limit = 200
	}
	var rows *sql.Rows
	var err error
	q := `SELECT job_id, filename, language, status, uploaded_at,
	             started_at, completed_at,
	             duration_seconds, processing_time_seconds, rtfx, confidence, error
	      FROM jobs`
	if status != "" {
		q += ` WHERE status=? ORDER BY uploaded_at DESC LIMIT ?`
		rows, err = s.db.Query(q, status, limit)
	} else {
		q += ` ORDER BY uploaded_at DESC LIMIT ?`
		rows, err = s.db.Query(q, limit)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var jobs []*job.Job
	for rows.Next() {
		j, err := scanJob(rows)
		if err != nil {
			return nil, err
		}
		jobs = append(jobs, j)
	}
	return jobs, rows.Err()
}

// Delete removes a job row. Caller is responsible for files.
func (s *Store) Delete(jobID string) error {
	_, err := s.db.Exec(`DELETE FROM jobs WHERE job_id=?`, jobID)
	return err
}

// Counts returns aggregate counts per status (for /healthz).
func (s *Store) Counts() (queued, running, done, failed int, err error) {
	rows, err := s.db.Query(`SELECT status, COUNT(*) FROM jobs GROUP BY status`)
	if err != nil {
		return 0, 0, 0, 0, err
	}
	defer rows.Close()
	for rows.Next() {
		var st string
		var n int
		if err := rows.Scan(&st, &n); err != nil {
			return 0, 0, 0, 0, err
		}
		switch st {
		case "queued":
			queued = n
		case "running":
			running = n
		case "done":
			done = n
		case "failed":
			failed = n
		}
	}
	return queued, running, done, failed, rows.Err()
}

// --- helpers ---

type rowScanner interface {
	Scan(dest ...any) error
}

func scanJob(r rowScanner) (*job.Job, error) {
	var (
		j        job.Job
		lang     sql.NullString
		started  sql.NullInt64
		complete sql.NullInt64
		dur      sql.NullFloat64
		proc     sql.NullFloat64
		rtfx     sql.NullFloat64
		conf     sql.NullFloat64
		errMsg   sql.NullString
		status   string
	)
	err := r.Scan(
		&j.JobID, &j.Filename, &lang, &status, &j.UploadedAt,
		&started, &complete, &dur, &proc, &rtfx, &conf, &errMsg,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	j.Status = job.Status(status)
	if lang.Valid {
		j.Language = lang.String
	}
	if started.Valid {
		j.StartedAt = started.Int64
	}
	if complete.Valid {
		j.CompletedAt = complete.Int64
	}
	if dur.Valid {
		j.DurationSeconds = dur.Float64
	}
	if proc.Valid {
		j.ProcessingTimeSeconds = proc.Float64
	}
	if rtfx.Valid {
		j.RTFx = rtfx.Float64
	}
	if conf.Valid {
		j.Confidence = conf.Float64
	}
	if errMsg.Valid {
		j.Error = errMsg.String
	}
	return &j, nil
}

func nullStr(s string) any {
	if s == "" {
		return nil
	}
	return s
}

File: modules/tools/denoiser-service/denoiser_service/store.py
"""SQLite-backed job persistence. Thread-safe (SQLite serializes writes; we use
a short-lived connection per call with WAL + busy_timeout).

Credentials are never written here — only the non-secret job metadata.
"""
from __future__ import annotations

import sqlite3
import time
from typing import Optional

from .models import Job, Status

_SCHEMA = """
CREATE TABLE IF NOT EXISTS jobs (
  job_id TEXT PRIMARY KEY,
  filename TEXT NOT NULL,
  source_url TEXT NOT NULL,
  dest_bucket TEXT NOT NULL,
  dest_key TEXT NOT NULL,
  status TEXT NOT NULL CHECK(status IN ('queued','running','done','failed')),
  uploaded_at INTEGER NOT NULL,
  started_at INTEGER,
  completed_at INTEGER,
  duration_seconds REAL,
  processing_time_seconds REAL,
  rtfx REAL,
  dest_url TEXT,
  error TEXT
);
CREATE INDEX IF NOT EXISTS idx_jobs_status ON jobs(status, uploaded_at);
"""

_COLS = (
    "job_id, filename, source_url, dest_bucket, dest_key, status, uploaded_at, "
    "started_at, completed_at, duration_seconds, processing_time_seconds, rtfx, "
    "dest_url, error"
)


def _now_ms() -> int:
    return int(time.time() * 1000)


class Store:
    def __init__(self, path: str):
        self._path = path
        with self._conn() as c:
            c.executescript(_SCHEMA)

    def _conn(self) -> sqlite3.Connection:
        c = sqlite3.connect(self._path, timeout=5.0)
        c.execute("PRAGMA journal_mode=WAL")
        c.execute("PRAGMA busy_timeout=5000")
        c.row_factory = sqlite3.Row
        return c

    def insert(self, j: Job) -> None:
        with self._conn() as c:
            c.execute(
                "INSERT INTO jobs (job_id, filename, source_url, dest_bucket, dest_key, "
                "status, uploaded_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
                (j.job_id, j.filename, j.source_url, j.dest_bucket, j.dest_key,
                 j.status.value, j.uploaded_at),
            )

    def insert_many(self, jobs: list[Job]) -> None:
        rows = [
            (j.job_id, j.filename, j.source_url, j.dest_bucket, j.dest_key,
             j.status.value, j.uploaded_at)
            for j in jobs
        ]
        with self._conn() as c:
            c.executemany(
                "INSERT INTO jobs (job_id, filename, source_url, dest_bucket, dest_key, "
                "status, uploaded_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
                rows,
            )

    def mark_running(self, job_id: str) -> None:
        with self._conn() as c:
            c.execute(
                "UPDATE jobs SET status='running', started_at=? "
                "WHERE job_id=? AND status IN ('queued','running')",
                (_now_ms(), job_id),
            )

    def mark_done(self, job_id: str, *, duration_s: float, proc_s: float,
                  rtfx: float, dest_url: str) -> None:
        with self._conn() as c:
            c.execute(
                "UPDATE jobs SET status='done', completed_at=?, duration_seconds=?, "
                "processing_time_seconds=?, rtfx=?, dest_url=? WHERE job_id=?",
                (_now_ms(), duration_s, proc_s, rtfx, dest_url, job_id),
            )

    def mark_failed(self, job_id: str, err: str) -> None:
        with self._conn() as c:
            c.execute(
                "UPDATE jobs SET status='failed', completed_at=?, error=? WHERE job_id=?",
                (_now_ms(), err, job_id),
            )

    def get(self, job_id: str) -> Optional[Job]:
        with self._conn() as c:
            row = c.execute(
                f"SELECT {_COLS} FROM jobs WHERE job_id=?", (job_id,)
            ).fetchone()
        return _row_to_job(row) if row else None

    def list(self, status: str = "", limit: int = 200) -> list[Job]:
        if limit <= 0 or limit > 1000:
            limit = 200
        with self._conn() as c:
            if status:
                rows = c.execute(
                    f"SELECT {_COLS} FROM jobs WHERE status=? "
                    "ORDER BY uploaded_at DESC LIMIT ?", (status, limit),
                ).fetchall()
            else:
                rows = c.execute(
                    f"SELECT {_COLS} FROM jobs ORDER BY uploaded_at DESC LIMIT ?", (limit,),
                ).fetchall()
        return [_row_to_job(r) for r in rows]

    def delete(self, job_id: str) -> None:
        with self._conn() as c:
            c.execute("DELETE FROM jobs WHERE job_id=?", (job_id,))

    def counts(self) -> dict[str, int]:
        out = {"queued": 0, "running": 0, "done": 0, "failed": 0}
        with self._conn() as c:
            for st, n in c.execute("SELECT status, COUNT(*) FROM jobs GROUP BY status"):
                out[st] = n
        return out

    def fail_orphans(self, msg: str) -> int:
        """At startup, any job left queued/running belongs to a previous process
        whose in-memory credentials are gone. Fail them with a clear message so
        the client can resubmit. Returns the number affected."""
        with self._conn() as c:
            cur = c.execute(
                "UPDATE jobs SET status='failed', completed_at=?, error=? "
                "WHERE status IN ('queued','running')",
                (_now_ms(), msg),
            )
            return cur.rowcount


def _row_to_job(r: sqlite3.Row) -> Job:
    return Job(
        job_id=r["job_id"],
        filename=r["filename"],
        source_url=r["source_url"],
        dest_bucket=r["dest_bucket"],
        dest_key=r["dest_key"],
        status=Status(r["status"]),
        uploaded_at=r["uploaded_at"],
        started_at=r["started_at"],
        completed_at=r["completed_at"],
        duration_seconds=r["duration_seconds"],
        processing_time_seconds=r["processing_time_seconds"],
        rtfx=r["rtfx"],
        dest_url=r["dest_url"],
        error=r["error"],
    )
